File: packages/flowise-attesta/src/AttestaGate.ts
import { INode, INodeData, INodeParams } from 'flowise-components';
import { DynamicTool } from '@langchain/core/tools';
import { Attesta, createActionContext, Verdict, RiskLevel } from '@attesta/core';
import type { ApprovalResult } from '@attesta/core';

class AttestaGate implements INode {
  label: string;
  name: string;
  version: number;
  type: string;
  icon: string;
  category: string;
  description: string;
  baseClasses: string[];
  inputs: INodeParams[];

  constructor() {
    this.label = 'Attesta Approval';
    this.name = 'attestaGate';
    this.version = 1.0;
    this.type = 'Tool';
    this.icon = 'attesta.svg';
    this.category = 'Tools';
    this.description = 'Human-in-the-loop approval for AI agent actions';
    this.baseClasses = ['Tool'];
    this.inputs = [
      {
        label: 'Function Name',
        name: 'functionName',
        type: 'string',
        placeholder: 'e.g. send_email, delete_record',
        description: 'Name of the action being gated',
      },
      {
        label: 'Risk Level',
        name: 'riskLevel',
        type: 'options',
        options: [
          { label: 'Auto (Score-based)', name: 'auto' },
          { label: 'Low', name: 'low' },
          { label: 'Medium', name: 'medium' },
          { label: 'High', name: 'high' },
          { label: 'Critical', name: 'critical' },
        ],
        default: 'auto',
        description: 'Risk level override. "Auto" uses the built-in risk scorer.',
      },
      {
        label: 'Risk Hints',
        name: 'riskHints',
        type: 'json',
        optional: true,
        additionalParams: true,
        default: '{}',
        description: 'JSON object of risk hints (e.g. {"destructive": true, "pii": true})',
      },
      {
        label: 'Tool Description',
        name: 'toolDescription',
        type: 'string',
        default: 'A gated action that requires approval before execution',
        description: 'Description shown to the LLM for this tool',
      },
    ];
  }

  async init(nodeData: INodeData): Promise<DynamicTool> {
    const functionName = (nodeData.inputs?.functionName as string) || 'gated_action';
    const riskLevelParam = (nodeData.inputs?.riskLevel as string) || 'auto';
    const riskHintsRaw = (nodeData.inputs?.riskHints as string) || '{}';
    const toolDescription = (nodeData.inputs?.toolDescription as string) || 'A gated action that requires approval before execution';

    let riskHints: Record<string, unknown> = {};
    try {
      riskHints = JSON.parse(riskHintsRaw);
    } catch {
      riskHints = {};
    }

    const riskOverride = riskLevelParam !== 'auto'
      ? riskLevelParam as typeof RiskLevel[keyof typeof RiskLevel]
      : undefined;

    const attesta = new Attesta({
      riskOverride,
      riskHints,
    });

    return new DynamicTool({
      name: functionName,
      description: toolDescription,
      func: async (input: string): Promise<string> => {
        let parsedInput: Record<string, unknown> = {};
        try {
          parsedInput = JSON.parse(input);
        } catch {
          parsedInput = { input };
        }

        const ctx = createActionContext({
          functionName,
          args: [],
          kwargs: parsedInput,
          hints: { ...riskHints },
          environment: 'production',
          metadata: { source: 'flowise' },
        });

        const result: ApprovalResult = await attesta.evaluate(ctx);

        if (
          result.verdict === Verdict.DENIED ||
          result.verdict === Verdict.TIMED_OUT ||
          result.verdict === Verdict.ESCALATED
        ) {
          return JSON.stringify({
            status: 'denied',
            verdict: result.verdict,
            riskScore: result.riskAssessment.score,
            riskLevel: result.riskAssessment.level,
            message: `Action "${functionName}" was denied by Attesta (risk: ${result.riskAssessment.level})`,
          });
        }

        return JSON.stringify({
          status: 'approved',
          verdict: result.verdict,
          riskScore: result.riskAssessment.score,
          riskLevel: result.riskAssessment.level,
          auditEntryId: result.auditEntryId,
          input: parsedInput,
        });
      },
    });
  }
}

module.exports = { nodeClass: AttestaGate };

File: packages/langflow-attesta/attesta_gate.py
"""Attesta Approval component for Langflow.

A human-in-the-loop approval component that evaluates AI agent actions for risk
before execution. This component integrates the Attesta framework into
Langflow workflows.

Contribution target: src/lfx/src/lfx/components/tools/attesta_gate.py
"""

from __future__ import annotations

import asyncio
import json
from typing import Any

try:
    from lfx.custom.custom_component.component import Component
    from lfx.io import DropdownInput, MessageTextInput, Output
    from lfx.schema import Data
except ImportError:
    # Lightweight local stubs so this module remains importable in OSS CI
    # without the full Langflow runtime installed.
    class Component:  # type: ignore[no-redef]
        def log(self, _message: str) -> None:
            return None

    class _InputBase:  # type: ignore[no-redef]
        def __init__(self, *args: Any, **kwargs: Any) -> None:
            self.args = args
            self.kwargs = kwargs

    class DropdownInput(_InputBase):  # type: ignore[no-redef]
        pass

    class MessageTextInput(_InputBase):  # type: ignore[no-redef]
        pass

    class Output(_InputBase):  # type: ignore[no-redef]
        pass

    class Data:  # type: ignore[no-redef]
        def __init__(self, data: dict[str, Any] | None = None) -> None:
            self.data = data or {}

from attesta.core.gate import Attesta
from attesta.core.types import ActionContext, RiskLevel, Verdict


class AttestaGate(Component):
    """Evaluates AI agent actions for risk and enforces approval policies."""

    display_name = "Attesta Approval"
    description = "Human-in-the-loop approval that evaluates AI agent actions for risk before execution"
    documentation = "https://attesta.dev"
    icon = "shield-check"
    name = "attesta_gate"

    inputs = [
        MessageTextInput(
            name="function_name",
            display_name="Function Name",
            info="Name of the action being gated (e.g. send_email, delete_record)",
            required=True,
        ),
        DropdownInput(
            name="risk_level",
            display_name="Risk Level",
            options=["auto", "low", "medium", "high", "critical"],
            value="auto",
            info="Risk level override. 'auto' uses the built-in risk scorer.",
        ),
        MessageTextInput(
            name="action_args",
            display_name="Action Arguments",
            info="JSON string of arguments to evaluate (e.g. {\"to\": \"user@example.com\"})",
            value="{}",
        ),
        MessageTextInput(
            name="risk_hints",
            display_name="Risk Hints",
            info="JSON string of risk hints (e.g. {\"destructive\": true, \"pii\": true})",
            value="{}",
            advanced=True,
        ),
    ]

    outputs = [
        Output(
            display_name="Approval Result",
            name="result",
            method="evaluate_gate",
        ),
    ]

    async def evaluate_gate(self) -> Data:
        """Evaluate the action through Attesta."""
        function_name: str = self.function_name
        risk_level_str: str = self.risk_level or "auto"

        # Parse JSON inputs
        action_args = self._parse_json(self.action_args, "action_args")
        risk_hints = self._parse_json(self.risk_hints, "risk_hints")

        # Determine risk override
        risk_override: RiskLevel | None = None
        if risk_level_str != "auto":
            risk_override = RiskLevel(risk_level_str)

        attesta = Attesta(
            risk_override=risk_override,
            risk_hints=risk_hints,
        )

        ctx = ActionContext(
            function_name=function_name,
            kwargs=action_args,
            hints=risk_hints,
            environment="production",
            metadata={"source": "langflow"},
        )

        result = await attesta.evaluate(ctx)

        denied = result.verdict in (
            Verdict.DENIED,
            Verdict.TIMED_OUT,
            Verdict.ESCALATED,
        )

        return Data(
            data={
                "verdict": result.verdict.value,
                "risk_score": result.risk_assessment.score,
                "risk_level": result.risk_assessment.level.value,
                "denied": denied,
                "audit_entry_id": result.audit_entry_id,
                "review_time_seconds": result.review_time_seconds,
                "function_name": function_name,
            }
        )

    def _parse_json(self, raw: str, field_name: str) -> dict[str, Any]:
        """Parse a JSON string input, returning {} on failure."""
        if not raw or raw.strip() == "":
            return {}
        try:
            parsed = json.loads(raw)
            if isinstance(parsed, dict):
                return parsed
            return {}
        except (json.JSONDecodeError, TypeError):
            self.log(f"Warning: invalid JSON in {field_name}, using empty dict")
            return {}
